### ISSUE TITLE
Cannot call pointerset at toplevel

### ISSUE THREAD

**Issue by timholy** [JuliaLang/julia#31182]
```julia
julia> TT = Union{UInt8, Int8}
Union{Int8, UInt8}

julia> a = TT[0x0, 0x1]
2-element Array{Union{Int8, UInt8},1}:
 0x00
 0x01

julia> pa = pointer(a)
Ptr{Union{Int8, UInt8}} @0x00007fc52a508bf0

julia> Base.pointerset(pa, 0x01, 2, 1)
ERROR: pointerset: invalid pointer
Stacktrace:
 [1] top-level scope at none:0
```
but
```julia
julia> unsafe_store!(pa, 0x01, 2)
Ptr{Union{Int8, UInt8}} @0x00007fc52a508bf0
```

Ref https://github.com/JuliaDebug/JuliaInterpreter.jl/issues/76

**Comment by timholy** [comment]
If this isn't trivially fixable, "calls to pointerset must be compiled" would be a clearer error message.

**Comment by Keno** [comment]
Seems like this should be disallowed in the compiled version also and whatever the users are of this should be fixed.

**Comment by timholy** [comment]
Comes from https://github.com/JuliaLang/julia/blob/4b01e6a60f750d1399cccd4fe447829cd75a2b2f/test/arrayops.jl#L2506-L2519

**Comment by JeffBezanson** [comment]
Yeah, that code doesn't even work:
```
julia> a = TT[0x0,0x0];

julia> pa = pointer(a);

julia> unsafe_store!(pa, 0x1, 2);

julia> a
2-element Array{Union{Int8, UInt8},1}:
 0x00
 0x00
```

**Comment by JeffBezanson** [comment]
I think we could allow it though, since we do have a defined memory layout for this. It would be a bit silly to allow this for references (we do, though of course there's a strong case against it) but not bits unions.

**Comment by vtjnash** [comment]
We can't entirely make it valid there—we have no possible way to know where the metadata for the value is located.

**Comment by JeffBezanson** [comment]
Right, it would not update the metadata. It would just be a way to manipulate the raw data representation. One can of course question the wisdom of that, but that's what the `unsafe` part is for :)
